flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/State.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.declarative;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.util.function.FunctionWithException;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import org.slf4j.Logger;
+
+import java.util.Optional;
+
+/**
+ * State abstraction of the {@link DeclarativeScheduler}. This interface contains all methods every
+ * state implementation must support.
+ */
+interface State {
```

Comment:
note: Had shared classes such as this been added first in a separate PR we could review all states in isolation without having to be based on one another.

Reply:
Good feedback, I didn't consider this. I didn't expect that much review on classes outside the test + I have never migrated a prototype to Flink, thus I need feedback like this (and I need to think harder beforehand myself ;) )